src/wallet/wallet.cpp

```diff
@@ -91,7 +91,45 @@ CPubKey CWallet::GenerateNewKey()
     bool fCompressed = CanSupportFeature(FEATURE_COMPRPUBKEY); // default to compressed public keys if we want 0.6.0 wallets
 
     CKey secret;
-    secret.MakeNewKey(fCompressed);
+
+    // Create new metadata
+    int64_t nCreationTime = GetTime();
+    CKeyMetadata metadata(nCreationTime);
+
+    // use HD key derivation if HD was enabled during wallet creation
+    if (!hdChain.masterKeyID.IsNull())
+    {
+        // for now we use a fixed keypath scheme of m/0'/0'/k
+        CKey key; //master key seed (256bit)
+        CExtKey masterKey; //hd master key
+        CExtKey accountKey; //key at m/0'
+        CExtKey externalChainChildKey; //key at m/0'/0'
+        CExtKey childKey; //key at m/0'/0'/<n>'
+
+        // try to get the master key
+        if (!GetKey(hdChain.masterKeyID, key))
+            throw std::runtime_error("CWallet::GenerateNewKey(): Master key not found");
+
+        assert(key.size() == 32);
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        masterKey.Derive(accountKey, 0 | 0x80000000);
+
+        // derive m/0'/0'
+        accountKey.Derive(externalChainChildKey, 0 | 0x80000000);
+
+        // derive child key at next index
+        externalChainChildKey.Derive(childKey, hdChain.nExternalChainCounter | 0x80000000);
```

Comment:
I added a loop to derive the next key as long as its not known by the wallet. I think this is required otherwise we might run into an exception if someone restores a backup or has imported a derived key of a higher child index.